### PR TITLE
Add `customTimers` option

### LINKED ISSUE #16: Feature request: accept custom `setTimeout` implementation
Hello. I am refactoring an existing codebase that uses a testing utility called [`sinon.useFakeTimers()`](https://sinonjs.org/releases/latest/fake-timers/). Calling this method basically replaces functions like `setTimeout` and `clearTimeout` in the global scope. I don't like this, but it's unfeasible to change this at this point.

The problem is: to work around another unrelated issue, I want to wrap some of these tests using `p-timeout`. Since `p-timeout` (obviously) uses `setTimeout` and `clearTimeout` from the global scope, which got replaced by `sinon.useFakeTimers()`, nothing times out correctly anymore. I would like to be able to make `p-timeout` keep using the original versions of `setTimeout` and `clearTimeout`.

Some ideas I had:

* Option 1: A new `customTimers` option could be added to allow the caller to specify custom `setTimeout` and `clearTimeout` implementations (and in my case, I would provide the original implementations to it). I would use it like this:
  ```js
  // my-module.js
  const { setTimeout, clearTimeout } = global;
  // ...
  await pTimeout(foo, 2000, undefined, { setTimeout, clearTimeout });
  ```
* Option 2: A new `preventTimerStubbing` option that, if true, would use the implementations of `setTimeout` and `clearTimeout` that were cached when the module was loaded (instead of reading the global scope directly), making it immune to future replacements of the global implementations. This has the downside that if `require('p-timeout')` is called too late, the replacements will already have taken place anyway. This wouldn't be a problem for my specific case though.
* Option 3: Just add `const { setTimeout, clearTimeout } = global` to the source code directly, making `p-timeout` immune to this type of mocking (unless the module is required too late, in which case there wouldn't be any difference). The downside is that it might break lots of people's codes, if they are also using `sinon.useFakeTimers()` and their code (or even any transitive dependency) uses `p-timeout` internally.

Would you be willing to accept any of these? If yes, which one? Let me know and I will open a PR. I prefer option 1, but it's your call.

Othewise, I understand - I might then just create and use a fork, or patch it with [`pirates`](https://www.npmjs.com/package/pirates).

Thank you!

### PR DESCRIPTION
**Pull request by papb**
Closes #16

### CODE REVIEW

File: index.test-d.ts [hunk 1]
```diff
@@ -23,5 +23,12 @@ pTimeout(delayedPromise(), 50, () => 10).then(value => {
 	expectType<string | number>(value);
 });
 
+pTimeout(delayedPromise(), 50, undefined, {customTimers: {setTimeout}});
+pTimeout(delayedPromise(), 50, 'foo', {customTimers: {clearTimeout}});
+pTimeout(delayedPromise(), 50, new Error('error'), {customTimers: {setTimeout, clearTimeout}});
+pTimeout(delayedPromise(), 50, () => 10, {});
+
+expectError(pTimeout(delayedPromise(), 50, () => 10, {customTimers: {setTimeout: () => true}}));
```
Comment (papb) [inline index.test-d.ts:31]: Interestingly, this call to `expectError` only worked after upgrading `tsd` from `0.7.2` to `0.13.1`.

**Comment by papb** [comment]
Hmm, looks like travis tests did not run? :thinking:

File: index.d.ts [hunk 2]
```diff
@@ -5,6 +5,39 @@ declare class TimeoutErrorClass extends Error {
...
+					setTimeout: originalSetTimeout,
+					clearTimeout: originalClearTimeout
+				}
+			});
+		})();
+		```
+		*/
+		readonly customTimers?: {
+			setTimeout?: typeof global.setTimeout;
+			clearTimeout?: typeof global.clearTimeout;
```
Comment (sindresorhus) [inline index.d.ts:38]: I don't think these should be optional. I think you should either have to set both or none.

**Comment by sindresorhus** [comment]
> Hmm, looks like travis tests did not run? 🤔

Travis stopped providing a free tier for open source... 

https://github.com/sindresorhus/p-timeout/commit/5e4f43cf37e156a4c8c9ea3280babae6b1dcbb91

**Comment by papb** [comment]
@sindresorhus I think you have to drop Node 8 support

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/p-timeout/releases/tag/v4.0.0

**Pull request merged**